### DISCUSSION TITLE
WS2812 and PWM at the same time?

### DISCUSSION THREAD

**Discussion by old-rotsch** [arendst/Tasmota#11658] [Q&A] [answered]
moin 
I'm building a wall clock with WS2812 LEDs. 
Question: can you configure WS2812 LEDs and 1-2 PWM outputs **at the same time**?
At the same time, I haven't got it to work yet.

**Comment by s-hadinger** [comment, marked as answer]
It's not supported. You could maybe control the pwm as channels and not light.

**Reply by quantenschaum** [reply]
How to do this? I can configure the module to use a WS2812 on one pin and another pin as PWM. The PWM does not appear in the web UI, but can I control it vie some commands? If then how?

**Reply by sfromis** [reply]
[SetOption15](https://tasmota.github.io/docs/Commands/#setoption15) allows you to select "basic PWM control", with pins controlled by the PWM command. Don't know if it works at the same time as neopixels, but it is a plausible thing to try.

**Reply by quantenschaum** [reply]
I does not seem to work. `pwm<x>` yields `RESULT = {"Command":"Error"}` when one pin is configured as WS2812, no matter if `SO15` is 1 or 0. `LedPwmMode` works, using `LedPwmOn` instead of `Pwm` (as workaround).

I would highly appreciate, if it was possible to combine a WS2813 strip with additional pwm channels (driving white strips). Or just to have a couple of independent sliders/channels using `MultiPwm`, 3 for the WS2812 and the others for the pwm pins.

**Reply by barbudor** [reply]
May be check 'setoption37 128'

**Reply by quantenschaum** [reply]
No, this does not work either. Using WS2812 seems to block all PWM functions completely.

**Comment by HoerMirAuf** [comment]
`LedPwmMode works, using LedPwmOn instead of Pwm (as workaround).`

Hi,

I know, it's a old Thread.
I want to use a WS2812 and a LED with PWM Dimming at the same time too.
But:
Configuring the GPIO to LED and setting to ledpwmmode 1, will result to fail colors an flickering on some LED's on the Neopixel WS2818.
Switching off ledpwmmode all went well with ne WS2812 neopixel

Is this a bug?

**Reply by ascillato** [reply]
Not a bug. As the selected answer explains, that combination is not supported in Tasmota. Sorry.

**Reply by sfromis** [reply]
With ESP32 (not ESP8266), you can have extra channels of both WS2812 and PWM, allowing you control outside of the regular light feature of Tasmota. I think that you'd be able to get what you want if configuring a PWM light the normal way, and have WS2812 channels above number 1, which you can control from Berry code only.

**Comment by HoerMirAuf** [comment]
Ok, thank you very much for your response.

Unfortunately I only have a ESP8266

What a pity. Thought it could be possible one year later ;)
... Unbelievable, something Tasmota can't do ;))

**Reply by sfromis** [reply]
You can get an ESP32 in the same form factor as the widespread ESP-12 modules, works for replacement in many situations, and not at all expensive.

**Comment by HoerMirAuf** [comment]
@sfromis .. thank you for this hint! Seems, a Wemos D1 Mini is replaceable with a D1 Mini ESP32.
Have ordered one to take a look.

**Reply by tasmotari** [reply]
Hi @HoerMirAuf,  do you have a success with ESP32 + Berry scripting? How hard to setup this?

**Comment by Basti149** [comment]
ESP32
Something has changed in the meantime. Is it now possible to switch a WS2812 Leds+ PWM outputs. And all of this via the GUI with a slider or is it only possible via the console?

Would it be practically possible to control WS2812 Leds+ 16 normal PWM outputs?

ESP12
WS2812+ with a single PWM output. is this possible now If I define the WS2812. Do I only see the three sliders or the single PWM output but not both (4 sliders).


ESP32
Hat sich in der zwischenzeit etwas geändert. Ist es jetzt möglich eine WS2812 Leds+ PWM Ausgänge zu schalten. Und das ganze noch über die GUI mit Schieberegler oder ist es nur per Console möglich?

Wäre es praktisch möglich WS2812 Leds+ 16 normale PWM Ausgänge zu steuern? 

ESP12 
WS2812+ mit einen einzelnen PWM Ausgang. Ist das jetzt möglich. Wenn ich die WS2812 definiere. Sehe ich nur die drei  Schieberegler oder den einzelnen PWM Ausgang aber nicht beides (4 Schieberegler).

**Reply by sfromis** [reply]
PWM is an entirely different way of controlling leds, controlling all leds on a strip at the same time. WS2812 is addressable leds (without anything PWM), where you can have individual control over each led on a strip. The basic light control in Tasmota can **only** handle one of either PWM or WS2812, not both types at the same time.

**Reply by s-hadinger** [reply]
The Web UI is necessarily limited. If you want to mix PWM and WS2812, it needs to be done via Berry scripting.

**Comment by ccaarrsstteenn** [comment]
I am also interested in using PWM and WS2812 in one project at the same time realzed with tasmota. Is it only a UI-Problem? Does anybody has a running Berry script for this? I am totally new to Berry.

**Reply by barbudor** [reply]
No need for Berry
You can by using `SetOption15 0`
You will be able to use the PWM output as a basic PWM instead of a Light using the PWM command
The DIMMER command would then address the WS2812

**Reply by sfromis** [reply]
No, unfortunately, that does not work.
The PWM command is not available with WS2812 + PWM pins configured, and `SetOption15 0`.

**Reply by sfromis** [reply]
If you config PWM with index 6 and above, in Berry you can address a PWM like this:
`gpio.set_pwm(gpio.pin(gpio.PWM1, 5), 1023)`
This is for PWM6, with max duty cycle of 1023. Minimum is 0 for off. Do not config PWM1-5.

You can also config WS2812 with index 2 and above, allowing you to control these via Berry (having the PWM channels as base Tasmota light feature), with greatly enhanced flexibility over the rudimentary base Tasmota support for neopixels, but then it gets to be some programming.
https://tasmota.github.io/docs/Berry_Addressable-LED/

**Reply by Lumpi-125** [reply]
Have you got any further with your Berry script? Ive been trying to unterstand, how the Berry Leds is working, but Im failing and failing...
Maybe somebody can give me a hint on my basic code:
I want to controll one PWM pin with tasmotas light controll and one WS2812 circle with berry.
I set the PWM pin as channel 1 and WS2812 pin as channel 2 in the module parameters. PWM is working as light.

Im trying to set the 1st Led to red.
Berry code:

var ring = Leds(12,  gpio.pin(gpio.WS2812, 2))  //WS2812 is connected to GPIO2 
ring.set_pixel_color(1, 0xFF0000)

Do i have to define the Leds class first? 

<img width="185" alt="image" src="https://user-images.githubusercontent.com/125072450/218505007-9e66f12b-5b96-4e19-a873-a991d95d2135.png">

**Reply by sfromis** [reply]
You need `ring.show()`, and your code addresses whatever gpio you have configured with WS2812 index 3 (or rather, not configured), trying to set color for the 2nd led.

Here's a working example for WS2812 index 2, a short strip of 8 pixels:
```
var pix = Leds(8, gpio.pin(gpio.WS2812, 1), Leds.WS2812_GRB)
pix.set_pixel_color(0, 0x004040)
pix.set_pixel_color(1, 0x400040)
pix.set_pixel_color(2, 0x404000)
pix.set_pixel_color(3, 0x404040)
pix.set_pixel_color(4, 0x000000)
pix.set_pixel_color(5, 0x000040)
pix.set_pixel_color(6, 0x004000)
pix.set_pixel_color(7, 0x400000)
pix.show()
```

Specifying `Leds.WS2812_GRB` was to be explicit, not really needed as this is the default. Indexes in Berry is generally relative to 0, while the gpio dropdown index starts with 1, hence dropdown index 2 being index 1 in the `gpio.pin` function call.

**Reply by Lumpi-125** [reply]
it works, thank you!

**Comment by Basti149** [comment]
Gibt es in der Zwischenzeit ein neuen Ansatz zusätzlich zu dem WS2812 Slider zwei PWM Ausgänge (weiß Töne) bei dem ESP32c3 hin zufügen. Leider ist es noch nicht möglich bei der v14.5

**Reply by sfromis** [reply]
The lights feature will be **either** WS2812 **or** PWM, not both at the same time.

However, you can control one of them via Berry (if you move to WS2812 index 2, or PWM index 6). PWM6 can also be controlled via the PWM command if reconfiguring.

**Comment by Basti149** [comment]
Thank you,

So I would have to set WS2812 1 to WS2812 2. And the two PWM outputs 1 and 2 (white). To PWM 5 and 6, right? This would allow me to switch the two PWM outputs via GUI and the WS2812 via Berry. Would it also be possible to switch the WS2812 via HTTP commands or MQTT?

Mit freundlichen Grüßen Sebastian
________________________________

**Reply by sfromis** [reply]
No, what I suggested would not be both, but either modifying WS2812 (as you posted now) **or** modify PWM, which would be index 6 and 7. However, as index 7 is not available on ESP32-C3, you cannot get both channels outside of the lights feature "owning" PWM1-5.

Hence your only option (for full operation on both types) would be modifying the neopixel WS2812, becoming controllable only via Berry code. You could then create your own control command (via Berry) available via MQTT etc.

**Comment by Basti149** [comment]
Vielen Dank da werde ich es mal probieren ich muss nicht unbedingt die WS2812 per Gui schalten da wären mir die beiden PWM-Ausgänge wichtiger. Wenn ich WS2812 als 2 einstelle dann könnte ich sie per MQTT ansteuern oder per Webbefehle (http://192.168.178.103/cm?cmnd=Led5 40,0,0<http://192.168.178.103/cm?cmnd=Led5%2040,0,0>) würde das funktionieren? Das würde mich bei den WS2812 schon zureichen?

Eine andere Frage, wenn ich ein ESP32S3 nehmen würde der Software mäßig mehr PWM-Ausgänge hätte könnte man dort PWM-Ausgänge und WS2812 gleichzeitig schalten und in der GUI anzeigen lassen?
________________________________
Von: sfromis ***@***.***>
Gesendet: Montag, 21. April 2025 12:44
An: arendst/Tasmota ***@***.***>
Cc: Smarthome ***@***.***>; Comment ***@***.***>
Betreff: Re: [arendst/Tasmota] WS2812 and PWM at the same time? (Discussion #11658)


No, what I suggested would not be both, but either modifying WS2812 (as you posted now) or modify PWM, which would be index 6 and 7. However, as index 7 is not available on ESP32-C3, you cannot get both channels outside of the lights feature "owning" PWM1-5.

Hence your only option (for full operation on both types) would be modifying the neopixel WS2812, becoming controllable only via Berry code. You could then create your own control command (via Berry) available via MQTT etc.

**Reply by sfromis** [reply]
Sure, if you leave out having any PWM, you do get normal operation with commands like `Led`.

But **again**, the lights feature of Tasmota can be **either** for PWM1-5 **or** WS2812-1, you cannot have both in use at the same time.

With Berry you can create control outside of the lights feature, including using Berry to add your own sliders. To do that with WS2812, that would mean coding the details of the handling.

**Comment by Basti149** [comment]
Vielen Dank zur Not müsste ich die beiden PWM-Ausgänge als Relais ändern oder über Scripting gehen.
Wäre es denn in Zukunft in Planung mit der Option beides zu aktivieren?
________________________________
Von: sfromis ***@***.***>
Gesendet: Montag, 21. April 2025 19:18
An: arendst/Tasmota ***@***.***>
Cc: Smarthome ***@***.***>; Comment ***@***.***>
Betreff: Re: [arendst/Tasmota] WS2812 and PWM at the same time? (Discussion #11658)


Sure, if you leave out having any PWM, you do get normal operation with commands like Led.

But again, the lights feature of Tasmota can be either for PWM1-5 or WS2812-1, you cannot have both in use at the same time.

With Berry you can create control outside of the lights feature, including using Berry to add your own sliders. To do that with WS2812, that would mean coding the details of the handling.

**Comment by DarthWeber** [comment]
I am trying to combine PWM and WS2812 but i did not understand the informations so far. My background: I want to use 6 PWM for servos only. I only use the PWM without shutter features. Additional i want to include a WS2812 which i want to control via MQTT and DDP only. I do not need any webcontrols, Is this possible? I use an esp32 wroom mini.

**Reply by sfromis** [reply]
If you want to use 6 PWM channels, a workable approach would be to **not** configure a WS2812 gpio, but use https://tasmota.github.io/docs/Berry_Addressable-LED/ to create a one-pixel strip. You could then create Berry commands for controlling this, but AFAIK this cannot work with the DDP support.

The other way would be to configure a gpio for WS2812 (allowing integration with DDP), but this does not combine with having PWM1-5 configured, only PWM numbers 6 and up. Number of PWM channels varies by ESP32 variant.

**Reply by DarthWeber** [reply]
okay. No worries. Will it work if i use WS2812 and a PCA9685 for PWM?

**Reply by sfromis** [reply]
Probably.

Still, it should already be workable if your specific ESP32 variant supports 11 PWM channels, allowing you to not use the first 5. IIRC, the "original" ESP32 provides up to 16 channels, which should be usable, but the newer variants tends to have fewer channels available.

What won't work is having gpios configured as WS2812 and PWM1-5 at the same time. That's why using PWM6 and up becomes an option, controlled via the `PWM` command or Berry coding.

**Reply by DarthWeber** [reply]
Thats the thing i don't understand. I use an esp32 wroom mini, according to tasmota and the template PWM1...16
<img width="369" height="970" alt="Screenshot 2026-01-14 120241" src="https://github.com/user-attachments/assets/013cf51f-3416-436e-b83e-ead687684344" />
but then the PWM6 command is unavailable and only reports "error" - independent of setoption15 (on or off)

**Reply by sfromis** [reply]
I may have gotten something mixed up, like PWM6 working outside of Tasmota lights support when not having a WS2812 configured.
It still works using the [gpio module](https://tasmota.github.io/docs/Berry/#gpio-module) in Berry. Like this to set PWM channel 6:
`gpio.set_pwm(gpio.pin(gpio.PWM1, 5), 123)`

Here's how you could use Berry to implement a command like `SetPWM6 456`:
```
tasmota.add_cmd('SetPWM',
  def (cmd, idx, payload)
    gpio.set_pwm(gpio.pin(gpio.PWM1, idx - 1), int(payload))
    tasmota.resp_cmnd_done()
  end)
```
It *can* also work without that `gpio.pin` indirection by directly hardcoding gpio number.

**Comment by Basti149** [comment]
Und zwar habe ich als indirekte Beleuchtung jeweils 1 Kanal weiß und 1 Kanal blaue LEDS die ich über PWM schalte und dimme zusätzlich möchte ich jedoch noch eine WS2812LED hinzufügen um meine Temperatur die ich per DS18b20 messe optisch zu visualisieren. Das wäre mein Wunsch aber leider kann ich zurzeit entweder das oder das hinzufügen.
________________________________
Von: DarthWeber ***@***.***>
Gesendet: Mittwoch, 14. Januar 2026 07:58
An: arendst/Tasmota ***@***.***>
Cc: Smarthome ***@***.***>; Comment ***@***.***>
Betreff: Re: [arendst/Tasmota] WS2812 and PWM at the same time? (Discussion #11658)


I am trying to combine PWM and WS2812 but i did not understand the informations so far. My background: I want to use 6 PWM for servos only. I only use the PWM without shutter features. Additional i want to include a WS2812 which i want to control via MQTT and DDP only. I do not need any webcontrols, Is this possible? I use an esp32 wroom mini.
